src/dxvk/dxvk_gpu_query.cpp

```diff
@@ -78,13 +73,44 @@ namespace dxvk {
 
 
   void DxvkGpuQuery::addQueryHandle(const DxvkGpuQueryHandle& handle) {
+    // Some apps forget to End() their queries; instead of accumulating query
+    // handles indefinitely, try to clean them up periodically.
+    if (m_handles.size() >= 256)
+      flushHandles();
+
     if (m_handle.queryPool)
       m_handles.push_back(m_handle);
     
     m_handle = handle;
   }
 
 
+  DxvkGpuQueryStatus DxvkGpuQuery::flushHandles() const {
+    // Consume handles in FIFO order, aggregating them into m_queryData
+    while (!m_handles.empty()) {
+      DxvkGpuQueryHandle handle = m_handles.front();
+
+      DxvkGpuQueryStatus status = getDataForHandle(m_queryData, handle);
+      if (status != DxvkGpuQueryStatus::Available)
+        return status;
+
+      handle.allocator->freeQuery(handle);
+      m_handles.pop_front();
+    }
+
+    if (m_handle.queryPool) {
```

Comment:
Probably easier to turn this whole mess into a `small_vector`. The main reason why there are two structures at all was to avoid allocations in the common case (1 query), but we have better ways to deal with this now.

Reply:
The concern that I have here is that the `small_vector`'s equivalent to `pop_front()` (`erase(0)`) is `O(n)`, which would make the while loop `O(n²)`. I suppose this could be mitigated if `small_vector::erase()` got a range overload, which could then be called once after the loop finishes; `flushHandles()` would be back to `O(n)`.

Either way, that's a separate change so I'd be rebasing the PR to add a commit if I do it.